Added initial DeepEval functionality to rag_answers evaluation

exfalsoquodlibet opened the pull request:
This PR introduces a structured framework for evaluating test cases using DeepEval. It includes the creation of data models for test cases, evaluation configurations, metrics and LLM as a Judge. Key features include:

- EvaluationTestCase: Represents test data, including questions, ideal answers, and retrieved contexts.
- EvaluationConfig: Configures metrics, LLM models, and evaluation settings.
- EvaluationResults: Captures and processes evaluation outcomes, providing methods for result serialisation and CSV/JSONL export.
- AggregatedResults: Summarises evaluation metrics, including per-input averages and statistics (mean, median, std).
- General evaluation workflow: A function (`evaluate_and_output_results`) orchestrates the evaluation process, integrates DeepEval, and exports results to CSV files.

I have also added:
- a possible structure where to save custom DeepEval metrics and LLM-as-a-Judge classes (to be added in a later PR)
- a `temp_example.py` to illustrate how it works together. I did not want to modify `rag_answers.cli.py` (which does fail now given these changes) before we have discussed this PR and agreed on its main components.

Pending:
I am adding unit tests...

Keen to discuss :)

kevindew commented:
Note: I've rebased this to take in the changes on main and rag-answers

kevindew commented on govuk_chat_evaluation/rag_answers/evaluate.py:
```diff
@@ -1,31 +1,227 @@
-        ],
-        threshold=0.5,
+    # initialise EvaluationResults with the config and the cases
+    evaluation_results = EvaluationResults(
```
I don't think this class name is quite right to reflect the purpose - as rather than being a representation of EvaluationResults it's more a mechanism to generate the results - so something like an EvaluationRunner perhaps?

I'd wonder if you did actually want this to be a class though and might want to create a module deepeval_evaluate.py that had an imported function into here. Producing code here of something like:
```
models = jsonl_to_models(evaluation_data_path, EvaluationTestCase)
evaluation_results = run_deepeval_evaluation(config=evaluation_config, cases=[model.to_llm_test_case() for model in models])
aggregation = AggregatedResults(evaluation_results)
```

exfalsoquodlibet replied:
Yes, I agree. I had a hard time trying to fit the orchestration into an `EvaluationResults` and `AggregationResults` class, following what is in the jailbreak guardrail, as I (mis)understood that was the general framework to respect.

I originally developed `run_deepeval_evaluation` to be a standalone function (https://github.com/alphagov/govuk-chat-evaluation-prototype/pull/18/commits/3fd49752a83497fc9d53efba515372c10e0d7a71).

Let's chat offline how to redesign this!

kevindew replied:
Ah sure, so for JailbreakGuardrails we have EvaluationResult singular rather than EvaluationResults, so that represents an individual check, whereas AggregateResults are the collective. So you have a one to many of one AggregateResults summarising multiple EvaluationResult objects. 

Yup let's find a time to chat - I imagine functions alone will suffice as I think it's mostly a procedure rather than something with state.

kevindew commented on govuk_chat_evaluation/rag_answers/evaluate.py:
```diff
@@ -1,31 +1,227 @@
+    _output_dir: Path, 
+    evaluation_data_path: Path, 
+    evaluation_config: EvaluationConfig, 
+    deepeval_evaluate_params: dict
```
I think rather than passing these through a few levels we'd do better to hardcode them initially as close to deepeval execute as we can and then if they become something we want to change have them be properties of the config object this accepts

kevindew replied:
Yeah sorry meant to comment on just that one line. 

Yeah I'd suggest getting them just hardcoded really close to where they're used and tweaking them directly. Otherwise it's a lot of passing of somewhat arbitrary data through the system.

kevindew commented on govuk_chat_evaluation/rag_answers/evaluate.py:
```diff
@@ -1,31 +1,227 @@
+        self.cases = cases
+        self.evaluation_results : list[TestResult] = []
+        
+    def run_deepeval_evaluation(self, **kwargs) -> None:
```
It seems a little surprising that this doesn't return the results but returns None - though I've made a suggestion a bit earlier around options to change the interface.

kevindew commented on govuk_chat_evaluation/rag_answers/evaluate.py:
```diff
@@ -1,31 +1,227 @@
+            **kwargs: Additional arguments to pass to the deepeval.evaluation function
+
+        Returns:
+            List of evaluation results for each model
```
Note: this doesn't match the function type signature.

kevindew commented on govuk_chat_evaluation/rag_answers/temp_example.py:
```diff
@@ -0,0 +1,46 @@
+        "temperature": 0.0
+    },
+    "n_runs": 3
+}
```
I had thought that the result of our discussion on how we would configure these was to be based on the individual metric so more like:

```
raw = {
    "metrics": [
        {"name": "faithfulness", "threshold": 0.8, "model": "gpt-4o", "temperature": 0.0, "n_runs": 3},
       {"name": "bias", "threshold": 0.5, "model": "gpt-4o", "temperature": 0.0, "n_runs": 3}
    ],
}
```

on the expectation that these may be changed per metric rather than be uniform?

exfalsoquodlibet replied:
Yes, you are right, and I initially started working based on that assumption. Forgot to add a comment about this when I opened the PR.

I then considered that:
- We should apply the same approach to the "wisdom of the crowd" (i.e., n_runs) consistently across all evaluation metrics. Although, one might argue that more complex metrics could justify additional runs.
- I was also concerned that our current setup for custom LLM-as-Judges might not work with DeepEval's asynchronous execution if we initialise multiple instances of the class.

Here as well, happy to chat offline.

kevindew commented on govuk_chat_evaluation/rag_answers/evaluate.py:
```diff
@@ -1,31 +1,225 @@
+                logger.info(f"Running evaluation iteration {i+1}/{self.n_runs}...")
+
+                evaluation_run = deepeval_evaluate(
+                    test_cases=cast(list[LLMTestCase | MLLMTestCase], self.cases), # hint to type checker: this is a List[LLMTestCase | MLLMTestCase], even though it's just List[LLMTestCase]
```
Nicely done, getting the type checker and deepeval to play nicely is such a pain: https://github.com/confident-ai/deepeval/pull/953

kevindew commented on govuk_chat_evaluation/rag_answers/evaluate.py:
```diff
@@ -1,31 +1,225 @@
+
+        logger.info("Deepval evaluation complete")
+        
+        self.evaluation_results = [results for run in all_evaluation_runs for results in run]
```
I imagine we might want to create a wrapper object for an individual evaluation result which can tie together each of the runs for a single question and be able to provide the individual metrics so that the aggregator object doesn't have the responsibility for working with the raw data as well as the aggregation.

kevindew replied:
Sure, might well be better to chat verbally like you suggest but I'll have a quick go via text and see if that helps progress.

So my expectation is that it would be handy to have an object that represents an individual question and the results associated with it so the wrapper would be at individual question level rather than a deepeval run.

As to how that might manifest:

I expect we'd want a class, say EvaluationResult, that stores perhaps the test case and the individual runs (or could be individual metric data if it went further - but that can probably be done by an abstraction. So something like:

```
# Temporary dict to build up the data
evaluation_runs = {test_case.input: { "test_case": test_case, runs: [] } } for test_case in test_cases}
for i in range(self.n_runs):
    evaluation_run = deepeval_evaluate(test_cases=self.cases, **kwargs)
    for result in evaluation_run.test_results:
      evaluation_runs[result.input]["runs"].append(result)

# Use our wrapper class to pull together the runs and the test case into a single object that can have individual methods for it's evaluation data.
# list[EvaluationResult]
evaluation_results = [EvaluationResult(test_case=run["test_case"], runs=run["runs"]) for run in evaluation_runs.values()]
```

exfalsoquodlibet replied:
Thanks Kev, that makes sense. Let's confirm when we talk, as I'd like to understand better why you think this is a more efficient approach.

kevindew commented on govuk_chat_evaluation/rag_answers/evaluate.py:
```diff
@@ -1,31 +1,225 @@
+        """Return the computed evaluation results."""
+        return self.evaluation_results
+    
+    @staticmethod
+    def _serialise_eval_result(result: TestResult) -> dict:
+        result_dict = asdict(result)
+        if "metrics_data" in result_dict:
+            result_dict["metrics_data"] = [
+                metric.model_dump() if isinstance(metric, BaseModel) else metric
+                for metric in result_dict["metrics_data"]
+            ]
+        return result_dict
+    
+    def to_serialised_dicts(self) -> list[dict[str, Any]]:
+        return [self._serialise_eval_result(result) for result in self.evaluation_results]
```
This aspect seems quite hard to understand - I can't find where this method is called but wondered if it's automatically called by something as part of serialisation?

This strikes me as sort of thing that could be a bit cleaner if the list of results was individual wrapper objects that had their own serialisation.

exfalsoquodlibet replied:
Thanks! Yeah, the idea here is to be able to save the raw evaluation results in a form that’s easy to reload and inspect later - especially if we want to dig into verbose logs or more detailed info. What is automatically saved by DeepEval is very hard to reload.

The default DeepEval TestResult objects aren't easily serialisable (e.g., for JSONL or CSV), so this method just flattens them into plain dictionaries that are easier to work with.

You're totally right though—it would probably be cleaner if each result were wrapped in an object that handled its own serialisation logic. That’s something we could refactor toward.

And good catch: I actually forgot to hook this up to the evaluate_and_output_results function so the serialised results get saved—will sort that!

exfalsoquodlibet replied:
I haven't acted on this yet. I just removed these methods in the meantime.

kevindew commented on govuk_chat_evaluation/rag_answers/evaluate.py:
```diff
@@ -1,31 +1,225 @@
+    def __init__(
+        self,
+        evaluation_results: list,
+        additional_metric_fields: list[str] = [],
+        additional_result_fields: list[str] = []
```
Shall we remove these until they are being used?

I found them a bit confusing myself and looked for any usage. I did wonder if rather than being additional they're more options (or even better would be that this class didn't need to know about them and could derive them from the results it's given).

exfalsoquodlibet replied:
Happy to remove them for now. They have only been needed for the `NonContradiction` metric but you are right, worth re-adding them in case only then in case.

kevindew commented on govuk_chat_evaluation/rag_answers/data_models.py:
```diff
@@ -10,11 +33,108 @@ class GenerateInput(BaseModel):
+    
+    @model_validator(mode="before")
+    @classmethod
+    def flatten_retrieved_context(cls, data: dict[str, Any]) -> dict[str, Any]:
```
I think rather than flattening this on input, it'd be a bit easier logically if retrieved_context was a list of StructuredContexts

Then on the StructuredContext model have a method of retrieved_context (or even __str__) which does the conversation.

kevindew commented on govuk_chat_evaluation/rag_answers/data_models.py:
```diff
@@ -1,6 +1,29 @@
+class GeneratedCaseParams(Enum):
+    QUESTION = "question"
+    LLM_ANSWER = "llm_answer"
+    RETRIEVED_CONTEXT = "retrieved_context"
```
This doesn't seem to be used anywhere?

exfalsoquodlibet replied:
No, you are right, I think it was from your initial implementation of the GEval class.

kevindew commented on govuk_chat_evaluation/rag_answers/data_models.py:
```diff
@@ -10,11 +33,108 @@ class GenerateInput(BaseModel):
+                return BiasMetric(threshold=self.threshold, model=llm_judge)
+
+
+class EvaluationConfig(BaseModel):
```
I am a bit confused about this and the related models. As per the comment below it sounds like you'd like to improve them too so perhaps we could have a chat about the individual needs and roles to iterate their modelling.

kevindew reviewed (commented):
Thanks Alessia, this looks great. Looks like there's lots of challenging bits involved in assembling this - so well done! I've put a bunch of comments in, happy to chat over them and/or prototype any ideas suggested further.

exfalsoquodlibet commented on .gitignore:
```diff
@@ -24,3 +24,6 @@ wheels/
 !/config/defaults
+
+# VSCode
+.vscode/
```
Thank you! I did not know about this convention, very useful.

I used to create a comprehensive .gitignore file using https://www.toptal.com/developers/gitignore/api/

Shall I remove the relevant commit from git history?

kevindew commented on govuk_chat_evaluation/rag_answers/evaluate.py:
```diff
@@ -1,31 +1,227 @@
+from .data_models import EvaluationTestCase, EvaluationConfig
+
+
+logger = logging.getLogger(__name__)
```
Sounds good - yeah it might well make sense to switch all the print statements to logger ones if our use case becomes more sophisticated. Wouldn't have objections to that, just would want us to avoid a mixture of both.

kevindew commented on govuk_chat_evaluation/timing.py:
```diff
@@ -0,0 +1,20 @@
+import time
+
+class Timer:
```
This class looks cool and very handy to use. 👍 

I was a little surprised that it prints a statement so I wondered it if it should be named accordingly and maybe converted to be function rather than a class, something like:

`with print_task_duration("Running Deepeval Evaluation")` 

in usage?

exfalsoquodlibet replied:
Yeah, I agree that the name could better signal the fact that it prints output as a side effect — especially since someone might assume it just records the time for later use. 

On the function vs class point: I went with a class because it leaves more flexibility like, if we later want to access .duration programmatically elsewhere, or extend it for logging instead of printing.
But I can turn this into a function via `contextlib` `@contextmanager` decorator.

kevindew commented on govuk_chat_evaluation/rag_answers/evaluate.py:
```diff
@@ -1,31 +1,225 @@
+
+        """
+
+        with Timer("Running Deepval Evaluation"):
+            logger.info("Running Deepval evaluation")
```
```suggestion
        with Timer("Running Deepeval Evaluation"):
            logger.info("Running Deepeval evaluation")
```

Couple of minor typos

kevindew commented on govuk_chat_evaluation/rag_answers/evaluate.py:
```diff
@@ -1,31 +1,123 @@
+            "std": mean_df.std(),
+        })
+
+    def export_to_csvs(self, _output_dir: Path, prefix: str = "metrics") -> None:
```
we should remove the underscore prefix from output_dir here as it's a used argument.

I don't think we really need the prefix argument. I see you've used it in your example one but in normal circumstance an output dir is created per run so it seems very unlikely we'd need to prefix it.

exfalsoquodlibet replied:
Good point. I also did not realise I had left an hardcoded prefix in the `evaluate_and_output_results` function... all fixed.

kevindew commented on govuk_chat_evaluation/rag_answers/data_models.py:
```diff
@@ -10,11 +38,102 @@ class GenerateInput(BaseModel):
+    metrics: list[MetricConfig]
+    n_runs: int
+
+    def get_metric_instances(self):
```
```suggestion
    def metric_instances(self):
```

nitpick that I don't think you should use a get_ prefix as it makes it look like a getter method and should have same meaning without that loaded prefix.

kevindew commented on govuk_chat_evaluation/rag_answers/data_models.py:
```diff
@@ -10,11 +38,102 @@ class GenerateInput(BaseModel):
+            case LLMJudgeModel.GEMINI_15_FLASH:
+                raise NotImplementedError(f"Judge model {self.model} instantiation not implemented.")
+            case LLMJudgeModel.GPT_4O_MINI | LLMJudgeModel.GPT_4O:
+                return self.model.value  # Just returns the string name as they are in-built models for DeepEval llm judge
```
Does this not make use of temperature?

kevindew commented on govuk_chat_evaluation/rag_answers/data_models.py:
```diff
@@ -10,11 +38,102 @@ class GenerateInput(BaseModel):
+                return self.model.value  # Just returns the string name as they are in-built models for DeepEval llm judge
+    
+
+class MetricConfig(BaseModel):
```
I think the creation of the LLM model instance could use some improvement (or even be deferred to a follow up PR as it's a bit a thorny). Where I think we're maybe at risk of twisting pydantic a bit beyond it's regular usage and things are a little complex.
 
I imagine things could be a bit simpler with something like:

```
def instantiate_llm_judge(model, temperature):
    match model:
         case LLMJudgeModel.AMAZON_NOVA_MICRO_1:
         # etc
        

class MetricConfig(BaseModel):
    name: MetricName
    threshold: float
    model: LLMJudgeModel
    temperature: float

    def to_metric_instance(self):
        model_instance = instantiate_llm_judge(model, temperature)
        match self.name:
            case MetricName.FAITHFULNESS:
                return FaithfulnessMetric(threshold=self.threshold, model=model)
            case MetricName.RELEVANCE:
                return AnswerRelevancyMetric(threshold=self.threshold, model=model)
            case MetricName.BIAS:
                return BiasMetric(threshold=self.threshold, model=model)
    
```

exfalsoquodlibet replied:
Happy to implement any change in another PR.

kevindew reviewed (commented):
Great stuff @exfalsoquodlibet - all looks to have much clearer domain concepts now. I've added another bunch of comments - most are on the minor side.

exfalsoquodlibet commented on govuk_chat_evaluation/rag_answers/evaluate.py:
```diff
@@ -1,31 +1,123 @@
 
 
+DEEPEVAL_EVAL_PARAMETERS = {
+    "print_results": "true",
```
You are right. Fixed.

exfalsoquodlibet commented:
Thanks for all your helpful comments and suggestions @kevindew . I hope addressed all of them. Let me know.

kevindew commented on govuk_chat_evaluation/rag_answers/evaluate.py:
```diff
@@ -1,31 +1,123 @@
+def evaluate_and_output_results(
+    output_dir: Path, 
+    evaluation_data_path: Path, 
+    evaluation_config: EvaluationConfig
```
I expect this won't be an instance of EvaluationConfig any more and would be an instance of Config 

(I also expect we'll have to move Config to be in the data_models.py file to avoid a cyclic dependency)

exfalsoquodlibet replied:
Implemented both these changes.

exfalsoquodlibet commented:
@kevindew I’ve implemented the changes we discussed before your leave:

Added unit tests across all relevant modules. The mocking was a bit tricky (especially around metric instantiation), so I’d appreciate a review to check if the logic and coverage make sense.

Refactored `cli.py` and moved the `Config` logic into `data_models.py`, removing the now-redundant `EvaluationConfig`.

I'm planning to handle the remaining items — specifically support for custom metrics and custom LLM judges — in a separate PR. I'd also like to add a better way to return and display evaluation results to make error analysis easier.

kevindew commented:
Thanks @exfalsoquodlibet, I did some work simplifying the tests on Thursday and I think, from my opinion, this should be good to merge into my branch and then into the repo, if you agree of course? I left all my changes in individual commits, happy to chat over them and/or rework them with you.

exfalsoquodlibet commented:
Thanks @kevindew! Happy to merge this into your branch. And also I'd be super keen if we could briefly chat over the main changes to the tests so that I can understand more how to best approach them a future time.
